BUG: fix read_gbq lost precision for longs above 2^53 and floats above 10k

tworec opened the pull request:
fixes:
- lost precision for longs above 2^53
- and floats above 10^4

Also contains  `test_gbq.py` clean up
- [x] closes #14020, #14305
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] updated docs
- [x] whatsnew entry

jreback commented on pandas/io/gbq.py:
```diff
@@ -553,11 +549,12 @@ def _parse_data(schema, rows):
         return float(field_value)
     elif field_type == 'TIMESTAMP':
-        timestamp = datetime.utcfromtimestamp(float(field_value))
-        return np.datetime64(timestamp)
```
why are you changing this?

tworec replied:
For two reasons:

1. it is independent from datetime lib
2. it is 4x faster

```
In [1]:
import timeit
timeit.timeit('import numpy as np; from datetime import datetime; np.datetime64(datetime.utcfromtimestamp(float(1234567890.123)))', number=1000000)
Out[1]:
6.163272747769952

In [2]:
timeit.timeit('import numpy as np; np.datetime64(int(float(1234567890.123)*1e6), "us")', number=1000000)
Out[2]:
1.5235848873853683
```

jreback replied:
1) doesn't really matter, but ok
2) sure, for a single value. you are much better off leaving these as floats, then coercing all in one go if you care about perf (using ``to_datetime``)

tworec replied:
It's not the point of this PR. I can restore previous version or leave mine - what do you prefer?

jreback commented on pandas/io/tests/test_gbq.py:
```diff
@@ -430,10 +426,16 @@ def test_should_properly_handle_valid_integers(self):
         query = 'SELECT INTEGER(NULL) as NULL_INTEGER'
         df = gbq.read_gbq(query, project_id=PROJECT_ID)
-        tm.assert_frame_equal(df, DataFrame({'NULL_INTEGER': [np.nan]}))
+        tm.assert_frame_equal(df, DataFrame({'NULL_INTEGER': [None]}))
```
this is not correct, ``np.nan`` is the missing entry

tworec replied:
np.NaN can not be stored in np.dtype(int)
http://pandas.pydata.org/pandas-docs/stable/gotchas.html#support-for-integer-na

jreback replied:
if u have missing values then by definition we cast to float always

we don't use None ever

jreback replied:
otherwise u end up with an object dtype which is generally not great

tworec replied:
casting to float means lost precision for integers > `2^53`.
I think it's better to lost performance than lost precision.
Lost precision means trouble when unnoticed.

I started with solution which raises exception for `NULL` in `INTEGER` column. This forces user to eliminate `NULL`-s (eg. by explicitly casting to `FLOAT` at BigQuery side; there is `float()` function). But finally I think it is less convenient.

My proposed solution is: 
- to store `INTEGER` column without `NULL`-s as `dtype = int64`
- and `INTEGER` with `NULL`-s as `dtype = object`

It mimics behaviour from `read_csv`.  Similarly for `BOOLEAN` columns (bool or object).

jreback replied:
this would need to be clearly documented.

tworec replied:
Ok, so you are ok with this approach?

jreback commented:
you are changing lots of things, but have minimal tests for this. pls add some.

jreback commented:
is this meant to close #14020 ?

tworec commented:
I didn't see #14020 before, but this PR seems to answer @aschmolck request

jreback commented on pandas/io/tests/test_gbq.py:
```diff
@@ -51,10 +51,6 @@ def _skip_if_no_private_key_contents():
                             "private key json contents")
 
-        _skip_if_no_project_id()
-        _skip_if_no_private_key_path()
```
pls restore

tworec replied:
it's a junk code. @parthea already removed these lines in f92cd7e

jreback commented:
needs more tests for datetimes

tworec commented:
this PR is all about numeric precision loss. I'll restore my datetime improvement. It can be added as separete PR I think.

tworec commented:
Ok, I've rebased and added full tests for this change (+ test clean up)

jorisvandenbossche commented:
cc @parthea

tworec commented:
@jreback, @parthea please review

jreback commented on doc/source/io.rst:
```diff
@@ -4428,16 +4428,11 @@ DataFrame with a shape and data types derived from the source table.
 to existing tables.
 
-You will need to install some additional dependencies:
-
```
is this not true anymore?

jreback replied:
then you need a pointer from the install.rst to the deps section

jreback replied:
you need to add a pointer from here to the Dependency section of the docs (that you added below).

jreback commented on doc/source/io.rst:
```diff
@@ -38,7 +38,7 @@ object.
     * :ref:`read_msgpack<io.msgpack>` (experimental)
     * :ref:`read_html<io.read_html>`
-    * :ref:`read_gbq<io.bigquery_reader>` (experimental)
+    * :ref:`read_gbq<io.bigquery>` (experimental)
```
why did you change this?

jreback commented on doc/source/io.rst:
```diff
@@ -4451,22 +4446,58 @@ The key functions are:
+Pandas supports these all `BigQuery data types <https://cloud.google.com/bigquery/data-types>`__:
+``STRING``, ``INTEGER`` (64bit), ``FLOAT`` (64 bit), ``BOOLEAN`` and
+``TIMESTAMP`` (microsecond precision). Data types ``BYTES`` and ``RECORD``
+are not supported.
```
are these unsupported types validated?

jreback commented on doc/source/io.rst:
```diff
@@ -4451,22 +4446,58 @@ The key functions are:
+are not supported.
+
+Integer and boolean ``NA`` handling
++++++++++++++++++++++++++++++++
```
needs to be the same length as the heading

jreback commented on doc/source/io.rst:
```diff
@@ -4451,22 +4446,58 @@ The key functions are:
+++++++++++++
+
+This module requires these additional dependencies:
+
```
ahh I c, ok well then make sure the pointer from the installation page goes here.

jreback commented on doc/source/io.rst:
```diff
@@ -4451,22 +4446,58 @@ The key functions are:
 Authentication
 ''''''''''''''
 
-.. versionadded:: 0.18.0
```
don't change things that are not related to this PR.

jreback commented on doc/source/io.rst:
```diff
@@ -4475,17 +4506,13 @@ is particularly useful when working on remote servers (eg. jupyter iPython noteb
 Additional information on
 `application default credentials <https://developers.google.com/identity/protocols/application-default-credentials>`__.
 
-.. versionadded:: 0.19.0
```
why are you taking things out like this? pls don't

jreback commented on doc/source/io.rst:
```diff
@@ -4451,22 +4446,58 @@ The key functions are:
+
+This is opposite to default pandas behaviour which will promote integer
+type to float in order to store NAs. See the :ref:`gotchas<gotchas.intna>`
+for detailed explaination.
```
acutally I disagree with this approach. I would *only* cast to ``object`` if the values are not representable (IOW big), otherwise follow the pandas standard and promot to ``float``. It is much more natural from a user point of view.

jreback replied:
@tworec we have 96 PR's open at the moment. I don't always comment till someone indicates things are ready.

can you *see* whether the column is nullable or not a-priori?

tworec replied:
ok, I understand :)

no, all columns in BigQuery queries are nullable

jreback commented on pandas/io/tests/test_gbq.py:
```diff
@@ -493,6 +497,39 @@ def test_should_read_as_service_account_with_key_contents(self):
         tm.assert_frame_equal(df, DataFrame({'VALID_STRING': ['PI']}))
 
+
+class TestReadGBQIntegrationWithServiceAccountKeyPath(tm.TestCase):
```
just make a new class in stead of changing existing code.

jorisvandenbossche commented on doc/source/whatsnew/v0.19.0.txt:
```diff
@@ -396,8 +396,9 @@ For ``MultiIndex``, values are dropped if any level is missing by default. Speci
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-The :func:`pandas.io.gbq.read_gbq` method has gained the ``dialect`` argument to allow users to specify whether to use BigQuery's legacy SQL or BigQuery's standard SQL. See the :ref:`docs <io.bigquery_reader>` for more details (:issue:`13615`).
-The :func:`pandas.io.gbq.to_gbq` method now allows the DataFrame column order to differ from the destination table schema (:issue:`11359`).
+- The :func:`pandas.io.gbq.read_gbq` method has gained the ``dialect`` argument to allow users to specify whether to use BigQuery's legacy SQL or BigQuery's standard SQL. See the :ref:`docs <io.bigquery_reader>` for more details (:issue:`13615`).
```
Or rather 0.20.0 ? As the current state of this PR changes the behaviour for certain integer columns, so I don't think it is needed to go into 0.19.1

jreback replied:
@jorisvandenbossche makes a good point, so 0.20.0

tworec replied:
ok

jreback commented:
I think it might be better to force the user to specify a ``dtype`` keyword to override / set behavior for particular columns (like we do in ``read_csv``; we sort of do this with ``coerce_float`` in ``read_sql``.

@jorisvandenbossche

tworec commented:
I offer solution with default behavior. 
Specifing `dtype` is cool, but it's new feature which I do not want to focus on right now. This PR is ment to fix things up, not to put new features.

Can we stop here?

tworec commented:
@jreback ping

jreback commented on doc/source/io.rst:
```diff
@@ -4452,7 +4447,43 @@ The key functions are:
+
+- `httplib2 <http://pypi.python.org/pypi/httplib2>`__
+- `google-api-python-client <http://github.com/google/google-api-python-client>`__
+- `oauth2client <https://github.com/google/oauth2client>`__.
```
is ``oauth2client`` optional?

tworec replied:
no, it is required.

tworec commented on doc/source/whatsnew/v0.19.1.txt:
```diff
@@ -0,0 +1,87 @@
+.. _whatsnew_0191:
+
```
sure

tworec commented on doc/source/whatsnew/v0.19.1.txt:
```diff
@@ -0,0 +1,87 @@
+Bug Fixes
+~~~~~~~~~
+
+- The :func:`pandas.io.gbq.to_gbq` method now stores ``INTEGER`` columns as ``dtype=object`` if they contain ``NULL`` values. Othewise they are stored as ``int64``. This prevents precision lost for integers greather than 2**53. Furthermore ``FLOAT`` columns with values above 10**4 are no more casted to ``int64`` which also caused precision lost (:issue: `14064`).
```
ok, I'm moving it to 0.20.0

tworec commented:
@jreback ok, I've rebased, moved to 0.20.0 and added install page ref to gbq deps

jorisvandenbossche commented on doc/source/io.rst:
```diff
@@ -4452,7 +4447,44 @@ The key functions are:
+Supported Data Types
+++++++++++++++++++++
+
+Pandas supports these all `BigQuery data types <https://cloud.google.com/bigquery/data-types>`__:
```
"these all" -> "all these" ?

tworec replied:
:) sorry, it's my polish 2 english calque

jorisvandenbossche commented on doc/source/io.rst:
```diff
@@ -4452,7 +4447,44 @@ The key functions are:
+Integer and boolean ``NA`` handling
++++++++++++++++++++++++++++++++++++
+
+.. versionadded:: 0.19
```
0.20

tworec replied:
ok

jorisvandenbossche commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -80,3 +80,5 @@ Performance Improvements
 Bug Fixes
 ~~~~~~~~~
+
+- The :func:`pandas.io.gbq.to_gbq` method now stores ``INTEGER`` columns as ``dtype=object`` if they contain ``NULL`` values. Othewise they are stored as ``int64``. This prevents precision lost for integers greather than 2**53. Furthermore ``FLOAT`` columns with values above 10**4 are no more casted to ``int64`` which also caused precision lost (:issue: `14064`).
```
`pandas.io.gbq.to_gbq` -> `~DataFrame.to_gbq`

tworec replied:
ok

jorisvandenbossche commented:
@tworec Added some minor doc comments. 
But, we also still have to decide about object vs float (https://github.com/pandas-dev/pandas/pull/14064#discussion_r82159392), I think? I find it a bit strange that we would take here a different approach than in all other places in pandas.

tworec commented:
@jorisvandenbossche I've corrected docs.

regarding float storage: my proposition is described in comment and implemented with tests
whats your proposition?

jreback commented:
@tworec sorry we got a bit lost in the shuffle.

can you rebase and we can get this in.

tworec commented:
@jreback ok, I've rebased

jreback commented:
@parthea thoughts?

tworec commented:
hi @jreback , can we finally merge this? Please decide if you want this change or drop it.

jreback commented:
I was waiting for @parthea to come back on this.

tworec commented:
oh, I c. We'll wait then. 
@parthea can you please check this out? :)

parthea commented on pandas/io/tests/test_gbq.py:
```diff
@@ -517,18 +554,72 @@ def test_should_properly_handle_valid_integers(self):
+        from math import pi
+        query = '''select * from
+                    (select PI() * POW(10, 307) as NULLABLE_DOUBLE),
+                    (select null as NULLABLE_DOUBLE)'''
```
SQL reserved words should be upper case for consistency with existing code.  We should either use the following or change existing code if there is a preference to change the SQL formatting.
```
        query = '''SELECT * FROM
                    (SELECT PI() * POW(10, 307) AS Nullable_Double),
                    (SELECT NULL AS Nullable_Double)'''
```

parthea commented on pandas/io/tests/test_gbq.py:
```diff
@@ -517,18 +554,72 @@ def test_should_properly_handle_valid_integers(self):
+        from math import pi
+        query = '''select * from
+                    (select PI() as NULLABLE_FLOAT),
+                    (select null as NULLABLE_FLOAT)'''
```
SQL keywords should be uppercase for consistency with existing code.

parthea commented on pandas/io/tests/test_gbq.py:
```diff
@@ -1124,7 +1228,7 @@ def setUpClass(cls):
+        _skip_local_auth_if_in_travis_env()
 
         _setup_common()
         clean_gbq_environment(_get_private_key_path())
```
Since we are testing local user account auth, can we change this to `clean_gbq_environment()`? We should also make the same change in tearDownClass.

parthea reviewed (commented):
I added a few minor comments. I think `_get_private_key_path()` should not be called in `TestToGBQIntegrationWithLocalUserAccountAuth`. 

Separately, it would be great if you could follow the steps in the [contributing docs to run the integration tests from your fork on Travis-CI](http://pandas.pydata.org/pandas-docs/stable/contributing.html#running-google-bigquery-integration-tests). You only need to do the Travis configuration described in the contributing docs once and you're all set for future PRs.

tworec reviewed (commented):
I fixed things as you @parthea suggested. Please review once again.

I don't fully trust travis CI so I need to create separete empty project for this purpose. And I'm not on my own. My company need to set up billing for it. It will take time.

tworec commented:
@parthea please review, because every time I look at this change it need manual rebase (conflicts).

tworec commented:
@parthea my Travis CI setup is broken, it causes errors with pickles eg.
`ERROR: pandas.io.tests.test_pickle.TestPickle.test_pickles('0.14.0',)`

https://travis-ci.org/RTBHOUSE/pandas/jobs/199354963

Can you help?

parthea commented:
@tworec I'm sorry for the delay. The latest commit looks good. 

Could you try running `git push --tags` on your master branch to see if that resolves the issue ?  Let me know if this works. I will submit a PR to update the BigQuery integration testing steps in the contributing docs to make it clear that you need to push tags to your master branch.

tworec commented:
@jreback thx

I've done it. Output is below. Why is this needed -- tags should not affect building right?
I guess this 141 commits is/was the problem. We'll see.

```
$ g co master 
Switched to branch 'master'
Your branch is ahead of 'origin/master' by 141 commits.
  (use "git push" to publish your local commits)

$ g ps --tags
To git@github.com:RTBHOUSE/pandas.git
 * [new tag]         v0.16.2 -> v0.16.2
 * [new tag]         v0.17.0 -> v0.17.0
 * [new tag]         v0.17.0rc1 -> v0.17.0rc1
 * [new tag]         v0.17.0rc2 -> v0.17.0rc2
 * [new tag]         v0.17.1 -> v0.17.1
 * [new tag]         v0.18.0 -> v0.18.0
 * [new tag]         v0.18.0rc1 -> v0.18.0rc1
 * [new tag]         v0.18.0rc2 -> v0.18.0rc2
 * [new tag]         v0.18.1 -> v0.18.1
 * [new tag]         v0.19.0 -> v0.19.0
 * [new tag]         v0.19.0rc1 -> v0.19.0rc1
 * [new tag]         v0.19.1 -> v0.19.1
 * [new tag]         v0.19.2 -> v0.19.2

```

Commit:
BUG: fix read_gbq lost numeric precision

fixes:
- lost precision for longs above 2^53
- and floats above 10k

tworec commented:
i've squashed

jreback commented:
@tworec when git clones a repo it doesn't clone the tags, so they are from *your* master whenever it was cloned, NOT the current one. This behavior baffles me, but that's how it works :<

tworec commented:
ok, now my travis build works well
thx for explanation

jreback commented:
thanks for the patience @tworec 

you can also put up a PR to: https://github.com/pydata/pandas-gbq (will need to take out the documentation changes, but outherwise should be clean).
